File: src/dev/xmodem_dw_uart.c
/*
 * xmodem_dw_uart.c
 *
 * Minimal XMODEM-1K receiver (CRC mode), adapted to use the provided
 * dw_uart_* API:
 *
 * Provided by your platform (already implemented by you):
 *   void dw_uart_putchar(char c);
 *   bool dw_uart_putchar_nb(char c);
 *   bool dw_uart_getchar_nb(char *c);
 *   bool dw_uart_rx_available(void);
 *   void dw_uart_flush(void);
 *
 * Required platform functions you must provide/implement:
 *   uint64_t platform_get_time_ms(void); // returns monotonic ms
 *
 * Optional (weak) helper to write to target memory; override if writing flash:
 *   void target_write(uint8_t *dst, const uint8_t *src, size_t len)
 *
 * Usage:
 *   ssize_t r = xmodem_receive_1k((uint8_t *)RECV_ADDR, RECV_MAXLEN);
 *   // if (r > 0) {  r bytes received at RECV_ADDR  }
 *
 * Host-side send command (Linux):
 *   stty -F /dev/ttyUSB0 115200 raw -echo cs8 -cstopb -parenb
 *   sx yourfile.bin < /dev/ttyUSB0 > /dev/ttyUSB0
 *
 */

#include "dev/t_dw_uart.h"
#include "t_types.h"
#include "lib/t_string.h"
#include "dev/t_timer.h"

/* control codes */
#define SOH    0x01
#define STX    0x02
#define EOT    0x04
#define ACK    0x06
#define NAK    0x15
#define CAN    0x18
#define CHAR_C 0x43

#define XMODEM_1K         1024
#define MAX_INIT_RETRIES  16
#define GETC_TIMEOUT_MS   10000  // 增加到 10 秒，给用户更多准备时间
#define MAX_BLOCK_RETRIES 16


/*
 * Platform time function (must be provided by your platform):
 *    uint64_t platform_get_time_ms(void);
 * It should return a monotonic millisecond counter used for timeouts.
 */
uint64_t
platform_get_time_ms(void)
{
    static uint64_t freq = 0;

    /* Lazy init frequency */
    if (freq == 0) {
        freq = CNTFRQ_EL0_READ();
        if (freq == 0) {
            /* Frequency read failed or zero - avoid divide by zero */
            return 0;
        }
    }

    uint64_t cnt = CNTPCT_EL0_READ();

#if 0
    /* (cnt * 1000) / freq computed in 128-bit to avoid overflow */
    __int128 tmp = (__int128) cnt * 1000;
    return (uint64_t) (tmp / (__int128) freq);
#else
    /*
     * Fallback without int128:
     * compute (cnt / freq) * 1000 + ((cnt % freq) * 1000) / freq
     * This keeps intermediate values within 64-bit.
     */
    uint64_t secs_part = cnt / freq;
    uint64_t rem       = cnt % freq;
    uint64_t ms        = secs_part * 1000;
    ms += (rem * 1000) / freq;
    return ms;
#endif
}

/* Optional target memory write helper (override if writing to flash) */
void
target_write(uint8_t *dst, const uint8_t *src, size_t len)
{
    memcpy(dst, src, len);
}

/* CRC16-CCITT (poly 0x1021), initial 0 */
static uint16_t
crc16_ccitt(const uint8_t *buf, size_t len)
{
    uint16_t crc = 0;
    while (len--) {
        crc ^= (uint16_t) (*buf++) << 8;
        for (int i = 0; i < 8; ++i) {
            if (crc & 0x8000)
                crc = (uint16_t) ((crc << 1) ^ 0x1021);
            else
                crc <<= 1;
        }
    }
    return crc;
}

/* helper: blocking put char (uses your blocking API) */
static inline void
uart_putc(uint8_t c)
{
    dw_uart_putchar((char) c);
}

/* helper: get a character with timeout (ms). returns -1 on timeout, 0..255 on success */
static int
uart_getc_timeout(int timeout_ms)
{
    uint64_t start    = platform_get_time_ms();
    uint64_t deadline = start + (uint64_t) timeout_ms;
    char     ch;
    while (platform_get_time_ms() <= deadline) {
        if (dw_uart_getchar_nb(&ch)) {
            return (int) (uint8_t) ch;
        }
        /* optionally yield or nop; if you have a light wait/sleep API you could call it here */
    }
    return -1;
}

/*
 * xmodem_receive_1k:
 *  - dst: target RAM address to write received file
 *  - maxlen: maximum number of bytes that can be stored at dst
 * returns number of bytes received (>0) or -1 on error
 */
ssize_t
xmodem_receive_1k(uint8_t *dst, size_t maxlen)
{
    uint8_t expected_blk = 1;
    size_t  write_offset = 0;
    int     c            = -1;
    int     init_try;

    /* Send initial 'C' (request CRC) several times until we see response */
    for (init_try = 0; init_try < MAX_INIT_RETRIES; ++init_try) {
        uart_putc(CHAR_C);
        c = uart_getc_timeout(GETC_TIMEOUT_MS);
        if (c >= 0)
            break;
    }
    if (c < 0) {
        /* no response from host */
        return -1;
    }

    /* main receive loop */
    for (;;) {
        if (c == EOT) {
            uart_putc(ACK);
            return (ssize_t) write_offset;
        } else if (c == CAN) {
            /* remote cancelled */
            return -1;
        } else if (c == SOH || c == STX) {
            size_t  block_size = (c == SOH) ? 128 : 1024;
            uint8_t header_blk, header_blk_comp;
            uint8_t block_buf[XMODEM_1K];
            uint8_t crc_hi, crc_lo;
            int     i;

            /* read block number and complement */
            int b1 = uart_getc_timeout(GETC_TIMEOUT_MS);
            int b2 = uart_getc_timeout(GETC_TIMEOUT_MS);
            if (b1 < 0 || b2 < 0) {
                uart_putc(NAK);
                return -1;
            }
            header_blk      = (uint8_t) b1;
            header_blk_comp = (uint8_t) b2;
            if ((uint8_t) (header_blk + header_blk_comp) != 0xFF) {
                uart_putc(NAK);
                c = uart_getc_timeout(GETC_TIMEOUT_MS);
                continue;
            }

            /* read data */
            for (i = 0; i < (int) block_size; ++i) {
                int ch = uart_getc_timeout(GETC_TIMEOUT_MS);
                if (ch < 0) {
                    uart_putc(NAK);
                    return -1;
                }
                block_buf[i] = (uint8_t) ch;
            }

            /* read CRC16 hi/lo */
            int hi = uart_getc_timeout(GETC_TIMEOUT_MS);
            int lo = uart_getc_timeout(GETC_TIMEOUT_MS);
            if (hi < 0 || lo < 0) {
                uart_putc(NAK);
                return -1;
            }
            crc_hi            = (uint8_t) hi;
            crc_lo            = (uint8_t) lo;
            uint16_t crc_recv = ((uint16_t) crc_hi << 8) | crc_lo;

            /* verify block number */
            if (header_blk != expected_blk) {
                /* duplicate of previous block? (sender didn't get our ACK) */
                if (header_blk == (uint8_t) (expected_blk - 1)) {
                    uart_putc(ACK); /* accept duplicate, do not write */
                    c = uart_getc_timeout(GETC_TIMEOUT_MS);
                    continue;
                } else {
                    uart_putc(NAK);
                    c = uart_getc_timeout(GETC_TIMEOUT_MS);
                    continue;
                }
            }

            /* verify CRC */
            uint16_t crc_calc = crc16_ccitt(block_buf, block_size);
            if (crc_calc != crc_recv) {
                uart_putc(NAK);
                c = uart_getc_timeout(GETC_TIMEOUT_MS);
                continue;
            }

            /* write to target memory (bounds check) */
            if (write_offset + block_size <= maxlen) {
                target_write(dst + write_offset, block_buf, block_size);
                write_offset += block_size;
            } else {
                /* overflow -> cancel transfer */
                uart_putc(CAN);
                uart_putc(CAN);
                return -1;
            }

            /* success for this block */
            uart_putc(ACK);
            expected_blk++;
            c = uart_getc_timeout(GETC_TIMEOUT_MS);
            continue;
        } else {
            /* unexpected char; try to read next */
            c = uart_getc_timeout(GETC_TIMEOUT_MS);
            if (c < 0) {
                /* timeout -> abort */
                uart_putc(CAN);
                uart_putc(CAN);
                return -1;
            }
        }
    }
}

// hexdump -C test.txt | head -n 2

// sudo minicom -D /dev/ttyUSB0 -b 1500000

// sudo stty -F /dev/ttyUSB0 1500000 raw -echo cs8 -cstopb -parenb

// sudo sh -c 'sx -k test.txt < /dev/ttyUSB0 > /dev/ttyUSB0'

File: src/dev/xmodem_dw_uart.rs
// XMODEM-1K receiver implementation in Rust
// Minimal XMODEM-1K receiver (CRC mode), adapted to use the dw_uart API
//
// This is a Rust port of xmodem_dw_uart.c with the same logic and structure

use crate::t_dw_uart::{dw_uart_getchar_nb, dw_uart_putchar};

// Control codes
const SOH: u8 = 0x01;
const STX: u8 = 0x02;
const EOT: u8 = 0x04;
const ACK: u8 = 0x06;
const NAK: u8 = 0x15;
const CAN: u8 = 0x18;
const CHAR_C: u8 = 0x43;

const XMODEM_1K: usize = 1024;
const MAX_INIT_RETRIES: i32 = 16;
const GETC_TIMEOUT_MS: u64 = 10000; // 10 seconds
const MAX_BLOCK_RETRIES: i32 = 16;

// Platform time function placeholder
// This should return a monotonic millisecond counter
// In the actual implementation, this would use the ARM system timer
fn platform_get_time_ms() -> u64 {
    // Placeholder implementation
    // In real code, read from CNTPCT_EL0 and CNTFRQ_EL0
    unsafe {
        let freq: u64;
        let cnt: u64;
        
        // Read counter frequency
        core::arch::asm!("mrs {}, cntfrq_el0", out(reg) freq);
        
        // Read counter value
        core::arch::asm!("mrs {}, cntpct_el0", out(reg) cnt);
        
        if freq == 0 {
            return 0;
        }
        
        // Calculate milliseconds
        let secs_part = cnt / freq;
        let rem = cnt % freq;
        let ms = secs_part * 1000;
        ms + (rem * 1000) / freq
    }
}

// Optional target memory write helper
// Override this if writing to flash or special memory regions
fn target_write(dst: &mut [u8], src: &[u8]) {
    dst[..src.len()].copy_from_slice(src);
}

// CRC16-CCITT (poly 0x1021), initial 0
fn crc16_ccitt(buf: &[u8]) -> u16 {
    let mut crc: u16 = 0;
    for &byte in buf {
        crc ^= (byte as u16) << 8;
        for _ in 0..8 {
            if (crc & 0x8000) != 0 {
                crc = (crc << 1) ^ 0x1021;
            } else {
                crc <<= 1;
            }
        }
    }
    crc
}

// Helper: blocking put char
#[inline(always)]
fn uart_putc(c: u8) {
    dw_uart_putchar(c);
}

// Helper: get a character with timeout (ms)
// Returns None on timeout, Some(byte) on success
fn uart_getc_timeout(timeout_ms: u64) -> Option<u8> {
    let start = platform_get_time_ms();
    let deadline = start + timeout_ms;
    
    while platform_get_time_ms() <= deadline {
        if let Some(ch) = dw_uart_getchar_nb() {
            return Some(ch);
        }
        // Yield or spin
        core::hint::spin_loop();
    }
    None
}

/// XMODEM-1K receiver
///
/// Receives a file via XMODEM-1K protocol into the provided buffer.
///
/// # Arguments
/// * `dst` - Target buffer to write received data
/// * `maxlen` - Maximum number of bytes that can be stored
///
/// # Returns
/// * `Ok(bytes_received)` - Number of bytes successfully received
/// * `Err(())` - Transfer failed or was cancelled
///
/// # Example
/// ```rust
/// let mut buffer = [0u8; 65536];
/// match xmodem_receive_1k(&mut buffer, buffer.len()) {
///     Ok(size) => println!("Received {} bytes", size),
///     Err(_) => println!("Transfer failed"),
/// }
/// ```
pub fn xmodem_receive_1k(dst: &mut [u8], maxlen: usize) -> Result<usize, ()> {
    let mut expected_blk: u8 = 1;
    let mut write_offset: usize = 0;
    
    // Send initial 'C' (request CRC) several times until we see response
    let mut c: Option<u8> = None;
    for _ in 0..MAX_INIT_RETRIES {
        uart_putc(CHAR_C);
        c = uart_getc_timeout(GETC_TIMEOUT_MS);
        if c.is_some() {
            break;
        }
    }
    
    let mut current_byte = match c {
        Some(byte) => byte,
        None => return Err(()), // No response from host
    };
    
    // Main receive loop
    loop {
        if current_byte == EOT {
            uart_putc(ACK);
            return Ok(write_offset);
        } else if current_byte == CAN {
            // Remote cancelled
            return Err(());
        } else if current_byte == SOH || current_byte == STX {
            let block_size = if current_byte == SOH { 128 } else { 1024 };
            
            // Read block number and complement
            let b1 = match uart_getc_timeout(GETC_TIMEOUT_MS) {
                Some(b) => b,
                None => {
                    uart_putc(NAK);
                    return Err(());
                }
            };
            
            let b2 = match uart_getc_timeout(GETC_TIMEOUT_MS) {
                Some(b) => b,
                None => {
                    uart_putc(NAK);
                    return Err(());
                }
            };
            
            let header_blk = b1;
            let header_blk_comp = b2;
            
            if header_blk.wrapping_add(header_blk_comp) != 0xFF {
                uart_putc(NAK);
                current_byte = match uart_getc_timeout(GETC_TIMEOUT_MS) {
                    Some(b) => b,
                    None => return Err(()),
                };
                continue;
            }
            
            // Read data block
            let mut block_buf = [0u8; XMODEM_1K];
            for i in 0..block_size {
                match uart_getc_timeout(GETC_TIMEOUT_MS) {
                    Some(ch) => block_buf[i] = ch,
                    None => {
                        uart_putc(NAK);
                        return Err(());
                    }
                }
            }
            
            // Read CRC16 hi/lo
            let crc_hi = match uart_getc_timeout(GETC_TIMEOUT_MS) {
                Some(b) => b,
                None => {
                    uart_putc(NAK);
                    return Err(());
                }
            };
            
            let crc_lo = match uart_getc_timeout(GETC_TIMEOUT_MS) {
                Some(b) => b,
                None => {
                    uart_putc(NAK);
                    return Err(());
                }
            };
            
            let crc_recv = ((crc_hi as u16) << 8) | (crc_lo as u16);
            
            // Verify block number
            if header_blk != expected_blk {
                // Duplicate of previous block? (sender didn't get our ACK)
                if header_blk == expected_blk.wrapping_sub(1) {
                    uart_putc(ACK); // Accept duplicate, do not write
                    current_byte = match uart_getc_timeout(GETC_TIMEOUT_MS) {
                        Some(b) => b,
                        None => return Err(()),
                    };
                    continue;
                } else {
                    uart_putc(NAK);
                    current_byte = match uart_getc_timeout(GETC_TIMEOUT_MS) {
                        Some(b) => b,
                        None => return Err(()),
                    };
                    continue;
                }
            }
            
            // Verify CRC
            let crc_calc = crc16_ccitt(&block_buf[..block_size]);
            if crc_calc != crc_recv {
                uart_putc(NAK);
                current_byte = match uart_getc_timeout(GETC_TIMEOUT_MS) {
                    Some(b) => b,
                    None => return Err(()),
                };
                continue;
            }
            
            // Write to target memory (bounds check)
            if write_offset + block_size <= maxlen && write_offset + block_size <= dst.len() {
                target_write(
                    &mut dst[write_offset..write_offset + block_size],
                    &block_buf[..block_size],
                );
                write_offset += block_size;
            } else {
                // Overflow -> cancel transfer
                uart_putc(CAN);
                uart_putc(CAN);
                return Err(());
            }
            
            // Success for this block
            uart_putc(ACK);
            expected_blk = expected_blk.wrapping_add(1);
            current_byte = match uart_getc_timeout(GETC_TIMEOUT_MS) {
                Some(b) => b,
                None => return Err(()),
            };
            continue;
        } else {
            // Unexpected char; try to read next
            current_byte = match uart_getc_timeout(GETC_TIMEOUT_MS) {
                Some(b) => b,
                None => {
                    // Timeout -> abort
                    uart_putc(CAN);
                    uart_putc(CAN);
                    return Err(());
                }
            };
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_crc16_ccitt() {
        // Test with known values
        let data = b"123456789";
        let crc = crc16_ccitt(data);
        // Expected CRC for "123456789" is 0x29B1
        assert_eq!(crc, 0x29B1);
    }

    #[test]
    fn test_crc16_ccitt_empty() {
        let data = b"";
        let crc = crc16_ccitt(data);
        assert_eq!(crc, 0);
    }
}
